wgengine/magicsock/magicsock.go

```diff
@@ -2468,8 +2386,68 @@ func (c *Conn) DERPs() int {
 	return len(c.activeDerp)
 }
 
-func (c *Conn) SetMark(value uint32) error { return nil }
-func (c *Conn) LastMark() uint32           { return 0 }
+// Bind returns the wireguard-go conn.Bind for c.
+func (c *Conn) Bind() conn.Bind {
+	return c.bind
+}
+
+// connBind is a wireguard-go conn.Bind for a Conn.
+type connBind struct {
+	*Conn
+	mu     sync.Mutex
+	closed bool
+}
+
+// Open is called by WireGuard to create a UDP binding.
+func (c *connBind) Open(uint16) ([]conn.ReceiveFunc, uint16, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if !c.closed {
+		return nil, 0, errors.New("magicsock: connBind already open")
+	}
+	c.closed = false
+
+	// Restore all receive calls.
+	c.pconn4.SetFakeClosed(false)
+	fns := []conn.ReceiveFunc{c.receiveIPv4, c.receiveDERP}
+	if c.pconn6 != nil {
+		c.pconn6.SetFakeClosed(false)
+		fns = append(fns, c.receiveIPv6)
+	}
+	// TODO: Combine receiveIPv4 and receiveIPv6 and receiveIP into a single
+	// closure that closes over a *RebindingUDPConn?
+	return fns, c.LocalPort(), nil
+}
+
+func (c *connBind) SetMark(value uint32) error {
+	return nil
+}
+
+func (c *connBind) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.closed {
+		return nil
+	}
+	c.closed = true
+
+	// Unblock all outstanding receives.
+	c.pconn4.SetFakeClosed(true)
+	if c.pconn6 != nil {
+		c.pconn6.SetFakeClosed(true)
+	}
+	// Send an empty read result to unblock receiveDERP,
+	// which will then check connBind.Closed.
+	c.derpRecvCh <- derpReadResult{}
```

Comment:
sends without a select on a closed channel always worry me. I've fought too many deadlocks from these.

Is there a closed channel we can select on? the derp reader's, the bind's, the magicsock.Conn's?

Reply:
There's not, really. The wireguard-go conn.Bind compact is that it calls the receive func repeatedly until it returns an error. The only context that matters is the lifetime of the wireguard-go device. But if we select on that, then we might leak a goroutine.

Reply:
But your comment did lead to a simplification and bug fix in processDERPReadResult, so thank you. :)